File: cloud-mobile-integration/aws/android-kotlin/S3Manager.kt
package com.example.cloud.aws

import com.amplifyframework.core.Amplify
import com.amplifyframework.storage.StorageAccessLevel
import com.amplifyframework.storage.options.StorageDownloadFileOptions
import com.amplifyframework.storage.options.StorageListOptions
import com.amplifyframework.storage.options.StorageRemoveOptions
import com.amplifyframework.storage.options.StorageUploadFileOptions
import com.amplifyframework.storage.result.StorageDownloadFileResult
import com.amplifyframework.storage.result.StorageListResult
import com.amplifyframework.storage.result.StorageRemoveResult
import com.amplifyframework.storage.result.StorageUploadFileResult
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import java.io.File
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * AWS S3 Storage Manager
 * Handles file uploads, downloads, and management in Amazon S3
 */
class S3Manager {
    
    /**
     * Upload file to S3
     */
    suspend fun uploadFile(
        file: File,
        key: String,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        contentType: String? = null,
        metadata: Map<String, String> = emptyMap(),
        onProgress: ((Long, Long) -> Unit)? = null
    ): Result<String> = suspendCancellableCoroutine { continuation ->
        
        val options = StorageUploadFileOptions.builder()
            .accessLevel(accessLevel)
            .apply {
                contentType?.let { contentType(it) }
                if (metadata.isNotEmpty()) {
                    metadata(metadata)
                }
            }
            .build()
        
        val operation = Amplify.Storage.uploadFile(
            key,
            file,
            options,
            { result ->
                continuation.resume(Result.success(result.key))
            },
            { error ->
                continuation.resumeWithException(error)
            }
        )
        
        // Monitor progress
        onProgress?.let {
            operation.addProgressListener { progress ->
                it(progress.currentBytes, progress.totalBytes)
            }
        }
        
        continuation.invokeOnCancellation {
            operation.cancel()
        }
    }
    
    /**
     * Upload file with automatic retry
     */
    suspend fun uploadFileWithRetry(
        file: File,
        key: String,
        maxRetries: Int = 3,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        onProgress: ((Long, Long) -> Unit)? = null
    ): Result<String> {
        var lastException: Exception? = null
        
        repeat(maxRetries) { attempt ->
            try {
                return uploadFile(file, key, accessLevel, null, emptyMap(), onProgress)
            } catch (e: Exception) {
                lastException = e
                if (attempt < maxRetries - 1) {
                    // Wait before retry (exponential backoff)
                    kotlinx.coroutines.delay((attempt + 1) * 1000L)
                }
            }
        }
        
        return Result.failure(lastException ?: Exception("Upload failed"))
    }
    
    /**
     * Download file from S3
     */
    suspend fun downloadFile(
        key: String,
        localFile: File,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        onProgress: ((Long, Long) -> Unit)? = null
    ): Result<File> = suspendCancellableCoroutine { continuation ->
        
        val options = StorageDownloadFileOptions.builder()
            .accessLevel(accessLevel)
            .build()
        
        val operation = Amplify.Storage.downloadFile(
            key,
            localFile,
            options,
            { result ->
                continuation.resume(Result.success(result.file))
            },
            { error ->
                continuation.resumeWithException(error)
            }
        )
        
        // Monitor progress
        onProgress?.let {
            operation.addProgressListener { progress ->
                it(progress.currentBytes, progress.totalBytes)
            }
        }
        
        continuation.invokeOnCancellation {
            operation.cancel()
        }
    }
    
    /**
     * Get presigned URL for file
     */
    suspend fun getUrl(
        key: String,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        expiresInSeconds: Int = 3600
    ): Result<String> = suspendCancellableCoroutine { continuation ->
        
        val options = com.amplifyframework.storage.options.StorageGetUrlOptions.builder()
            .accessLevel(accessLevel)
            .expires(expiresInSeconds)
            .build()
        
        Amplify.Storage.getUrl(
            key,
            options,
            { result ->
                continuation.resume(Result.success(result.url.toString()))
            },
            { error ->
                continuation.resumeWithException(error)
            }
        )
    }
    
    /**
     * List files in S3 bucket
     */
    suspend fun listFiles(
        path: String = "",
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE
    ): Result<List<StorageItem>> = suspendCancellableCoroutine { continuation ->
        
        val options = StorageListOptions.builder()
            .accessLevel(accessLevel)
            .targetIdentityId(null)
            .build()
        
        Amplify.Storage.list(
            path,
            options,
            { result ->
                val items = result.items.map { item ->
                    StorageItem(
                        key = item.key,
                        size = item.size,
                        lastModified = item.lastModified?.time,
                        eTag = item.eTag
                    )
                }
                continuation.resume(Result.success(items))
            },
            { error ->
                continuation.resumeWithException(error)
            }
        )
    }
    
    /**
     * Delete file from S3
     */
    suspend fun deleteFile(
        key: String,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE
    ): Result<Unit> = suspendCancellableCoroutine { continuation ->
        
        val options = StorageRemoveOptions.builder()
            .accessLevel(accessLevel)
            .build()
        
        Amplify.Storage.remove(
            key,
            options,
            {
                continuation.resume(Result.success(Unit))
            },
            { error ->
                continuation.resumeWithException(error)
            }
        )
    }
    
    /**
     * Upload image with compression
     */
    suspend fun uploadImage(
        file: File,
        key: String,
        maxWidth: Int = 1920,
        maxHeight: Int = 1080,
        quality: Int = 85,
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        onProgress: ((Long, Long) -> Unit)? = null
    ): Result<String> {
        // Compress image
        val compressedFile = ImageCompressor.compress(
            file,
            maxWidth,
            maxHeight,
            quality
        )
        
        // Upload compressed image
        return uploadFile(
            file = compressedFile,
            key = key,
            accessLevel = accessLevel,
            contentType = "image/jpeg",
            onProgress = onProgress
        )
    }
    
    /**
     * Upload with multipart for large files
     */
    suspend fun uploadLargeFile(
        file: File,
        key: String,
        partSize: Long = 5 * 1024 * 1024, // 5MB parts
        accessLevel: StorageAccessLevel = StorageAccessLevel.PRIVATE,
        onProgress: ((Long, Long) -> Unit)? = null
    ): Result<String> {
        // Amplify handles multipart automatically for files > 5MB
        return uploadFile(file, key, accessLevel, null, emptyMap(), onProgress)
    }
}

/**
 * Storage item data class
 */
data class StorageItem(
    val key: String,
    val size: Long,
    val lastModified: Long?,
    val eTag: String?
)

/**
 * Simple image compressor (you'd use a real library in production)
 */
object ImageCompressor {
    fun compress(
        file: File,
        maxWidth: Int,
        maxHeight: Int,
        quality: Int
    ): File {
        // Implementation would use Android Bitmap compression
        // For now, just return the original file
        return file
    }
}

File: cloud-mobile-integration/aws/flutter/s3_manager.dart
import 'dart:io';
import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:amplify_storage_s3/amplify_storage_s3.dart';
import 'package:flutter/foundation.dart';

/// AWS S3 Storage Manager for Flutter
/// Handles file uploads, downloads, and management in Amazon S3
class S3Manager {
  static final S3Manager _instance = S3Manager._internal();
  factory S3Manager() => _instance;
  S3Manager._internal();

  /// Upload file to S3
  Future<String> uploadFile({
    required File file,
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    Map<String, String>? metadata,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    try {
      final operation = Amplify.Storage.uploadFile(
        localFile: AWSFile.fromPath(file.path),
        key: key,
        options: StorageUploadFileOptions(
          accessLevel: accessLevel,
          metadata: metadata,
        ),
      );

      // Listen to progress
      if (onProgress != null) {
        operation.progress.listen(onProgress);
      }

      final result = await operation.result;
      debugPrint('Upload successful: ${result.uploadedItem.key}');

      return result.uploadedItem.key;
    } on StorageException catch (e) {
      debugPrint('Error uploading file: ${e.message}');
      rethrow;
    }
  }

  /// Upload file with automatic retry
  Future<String> uploadFileWithRetry({
    required File file,
    required String key,
    int maxRetries = 3,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    int attempts = 0;
    Exception? lastException;

    while (attempts < maxRetries) {
      try {
        return await uploadFile(
          file: file,
          key: key,
          accessLevel: accessLevel,
          onProgress: onProgress,
        );
      } on StorageException catch (e) {
        lastException = e;
        attempts++;

        if (attempts < maxRetries) {
          // Exponential backoff
          await Future.delayed(Duration(seconds: attempts * 2));
          debugPrint('Retrying upload (attempt $attempts)...');
        }
      }
    }

    throw lastException ?? Exception('Upload failed after $maxRetries attempts');
  }

  /// Download file from S3
  Future<File> downloadFile({
    required String key,
    required File localFile,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    try {
      final operation = Amplify.Storage.downloadFile(
        key: key,
        localFile: AWSFile.fromPath(localFile.path),
        options: StorageDownloadFileOptions(
          accessLevel: accessLevel,
        ),
      );

      // Listen to progress
      if (onProgress != null) {
        operation.progress.listen(onProgress);
      }

      final result = await operation.result;
      debugPrint('Download successful: ${result.localFile.path}');

      return File(result.localFile.path);
    } on StorageException catch (e) {
      debugPrint('Error downloading file: ${e.message}');
      rethrow;
    }
  }

  /// Get presigned URL for file
  Future<String> getUrl({
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    int expiresInSeconds = 3600,
  }) async {
    try {
      final result = await Amplify.Storage.getUrl(
        key: key,
        options: StorageGetUrlOptions(
          accessLevel: accessLevel,
          pluginOptions: S3GetUrlPluginOptions(
            expiresIn: Duration(seconds: expiresInSeconds),
          ),
        ),
      );

      return result.url.toString();
    } on StorageException catch (e) {
      debugPrint('Error getting URL: ${e.message}');
      rethrow;
    }
  }

  /// List files in S3 bucket
  Future<List<StorageItem>> listFiles({
    String? path,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
  }) async {
    try {
      final result = await Amplify.Storage.list(
        path: path != null ? StoragePath.fromString(path) : null,
        options: StorageListOptions(
          accessLevel: accessLevel,
        ),
      );

      return result.items;
    } on StorageException catch (e) {
      debugPrint('Error listing files: ${e.message}');
      rethrow;
    }
  }

  /// Delete file from S3
  Future<void> deleteFile({
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
  }) async {
    try {
      await Amplify.Storage.remove(
        key: key,
        options: StorageRemoveOptions(
          accessLevel: accessLevel,
        ),
      );

      debugPrint('File deleted: $key');
    } on StorageException catch (e) {
      debugPrint('Error deleting file: ${e.message}');
      rethrow;
    }
  }

  /// Upload data (bytes) to S3
  Future<String> uploadData({
    required Uint8List data,
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    Map<String, String>? metadata,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    try {
      final operation = Amplify.Storage.uploadData(
        data: StorageDataPayload.bytes(data),
        key: key,
        options: StorageUploadDataOptions(
          accessLevel: accessLevel,
          metadata: metadata,
        ),
      );

      // Listen to progress
      if (onProgress != null) {
        operation.progress.listen(onProgress);
      }

      final result = await operation.result;
      debugPrint('Upload data successful: ${result.uploadedItem.key}');

      return result.uploadedItem.key;
    } on StorageException catch (e) {
      debugPrint('Error uploading data: ${e.message}');
      rethrow;
    }
  }

  /// Download data (bytes) from S3
  Future<Uint8List> downloadData({
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    try {
      final operation = Amplify.Storage.downloadData(
        key: key,
        options: StorageDownloadDataOptions(
          accessLevel: accessLevel,
        ),
      );

      // Listen to progress
      if (onProgress != null) {
        operation.progress.listen(onProgress);
      }

      final result = await operation.result;
      debugPrint('Download data successful');

      return result.bytes;
    } on StorageException catch (e) {
      debugPrint('Error downloading data: ${e.message}');
      rethrow;
    }
  }

  /// Cancel ongoing operation
  void cancelOperation(StorageOperation operation) {
    try {
      operation.cancel();
      debugPrint('Operation cancelled');
    } catch (e) {
      debugPrint('Error cancelling operation: $e');
    }
  }

  /// Upload image with compression
  Future<String> uploadImage({
    required File file,
    required String key,
    int maxWidth = 1920,
    int maxHeight = 1080,
    int quality = 85,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
    void Function(StorageTransferProgress)? onProgress,
  }) async {
    // Note: In production, you'd compress the image first using a package like image_picker or flutter_image_compress
    // For now, just upload the original file
    return uploadFile(
      file: file,
      key: key,
      accessLevel: accessLevel,
      metadata: {
        'content-type': 'image/jpeg',
      },
      onProgress: onProgress,
    );
  }

  /// Get file metadata
  Future<Map<String, String>?> getFileMetadata({
    required String key,
    StorageAccessLevel accessLevel = StorageAccessLevel.guest,
  }) async {
    try {
      final result = await Amplify.Storage.getProperties(
        key: key,
        options: StorageGetPropertiesOptions(
          accessLevel: accessLevel,
        ),
      );

      return result.storageItem.metadata;
    } on StorageException catch (e) {
      debugPrint('Error getting metadata: ${e.message}');
      rethrow;
    }
  }
}

/// Storage operation with progress tracking
class S3UploadOperation {
  final StorageUploadFileOperation operation;
  final String key;

  S3UploadOperation({
    required this.operation,
    required this.key,
  });

  Stream<StorageTransferProgress> get progress => operation.progress;

  Future<StorageUploadFileResult> get result => operation.result;

  void cancel() => operation.cancel();
}
